### DISCUSSION TITLE
v3 Different Themes per Tenant

### DISCUSSION THREAD

**Discussion by claybitner** [filamentphp/filament#7826] [Help] [answered]
I have a color_theme table with a few variables such as primary_color, and each tenant has a color_theme associated with it. I'm trying to modify the theme for my tenants based on that color. Before the theme is set, I can't seem to access the current tenant. I'm trying to set 
```
$panel->colors([
            'primary' => $tenant->color_theme->primary_color,
        ]);

```
Even though I've tried to set $tenant with either `$panel->hasTenancy();` or `$tenant = Filament::getTenant();`, it can't seem to access the actual tenant that they're using.

Am I missing something obvious?

**Comment by pboivin** [comment]
I think you can do this in a service provider (e.g. AppServiceProvider):
```php
public function boot(): void
{
    Filament::serving(function () {
        // get $tenant
        
        FilamentColor::register(['primary' => $tenant->color_theme->primary_color]);
    });
}
```

**Reply by claybitner** [reply]
Tried this:
```
public function boot(): void
    {
        Filament::serving(function () {
            $tenant = Filament::getTenant();
            if ($tenant && $tenant->color_theme) {
                FilamentColor::register(['primary' => $tenant->color_theme->primary_color]);
            }
        });
    }
```
But `$tenant` is always null. Even when I'm on a page that has selected a tenant. I guess I'm not understanding at what point the tenant gets injected.

**Reply by pboivin** [reply]
Ah, makes sense... I think the `Filament::serving()` callbacks run before the tenant is identified. Not really sure at this point, but it seems like you could do it as part of the Panel boot instead:

```php
class AdminPanelProvider extends PanelProvider
{
    public function panel(Panel $panel): Panel
    {
        return $panel
            // ...
            ->bootUsing(function () {
                $tenant = Filament::getTenant();
                if ($tenant && $tenant->color_theme) {
                    FilamentColor::register(['primary' => $tenant->color_theme->primary_color]);
                }
            });
```

**Comment by claybitner** [comment, marked as answer]
So it turns out the solution is to actually create a Middleware for it.

I created a Middleware called, `ApplyTenantThemeColors` with the handle function like this:
```php
    public function handle(Request $request, Closure $next): Response
    {
        $tenant = Filament::getTenant();
        if ($tenant && $tenant->color_theme) {
            FilamentColor::register(['primary' => $tenant->color_theme->primary_color]);
        }
        return $next($request);
    }

```

Then inside the Panel Provider, I added the tenantMiddleware:
```php
            ->tenantMiddleware([
                ApplyTenantThemeColors::class,
            ])
```

**Reply by pboivin** [reply]
Nice, thanks for sharing!

**Comment by coolsam726** [comment]
I was encountering the exact same problem. The middleware solution is perfect. Additionally, I am thinking of doing a PR to allow closures in the `$panel->colors()` method which will allow us to do something like this:

```php
$panel->colors(['primary' => fn() => tenant()->primary_color]);
```
or

```php
FilamentColor::register(['primary' => fn() => tenant()->primary_color]);
```
I have already tested it locally and it works perfectly.
